org.eclipse.xtext/src/org/eclipse/xtext/resource/persistence/ResourceStorageLoadable.java

```diff
@@ -99,21 +102,25 @@ protected void readResourceDescription(StorageAwareResource resource, InputStrea
 	}
 
 	protected void readNodeModel(StorageAwareResource resource, InputStream inputStream) throws IOException {
-		SerializableNodeModel serializableNodeModel = new SerializableNodeModel(resource);
-		// if this is a synthetic resource (i.e. tests or so, don't load the node model)
-		if (!resource.getResourceSet().getURIConverter().exists(resource.getURI(),
-				resource.getResourceSet().getLoadOptions())) {
-			LOG.info("Skipping loading node model for synthetic resource " + resource.getURI());
-			return;
-		}
-		try (InputStreamReader reader = new InputStreamReader(resource.getResourceSet().getURIConverter().createInputStream(resource.getURI()),
-				resource.getEncoding())) {
-			String completeContent = CharStreams.toString(reader);
-			DeserializationConversionContext deserializationContext = new DeserializationConversionContext(resource,
-					completeContent);
-			serializableNodeModel.readObjectData(new DataInputStream(inputStream), deserializationContext);
-			resource.setParseResult(new ParseResult(head(resource.getContents()), serializableNodeModel.root,
-					deserializationContext.hasErrors()));
+		if (!resource.customReadNodeModel(inputStream)) {
```

Comment:
Resources can opt-in to handle the serialization themselves. The rest is mostly changed indentation and extracting local vars.